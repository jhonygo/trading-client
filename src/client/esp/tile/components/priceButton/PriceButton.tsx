import React from 'react';
import classNames from 'classnames';
import { TradeRequest } from '../../model/tradeRequest';
import * as style from './style.css';
import styled from 'styled-components';
import { Price } from '../price/Price';

export const TileStaleStyle = styled.div`
  position: absolute;
  height: 100%;
  width: 100%;
  border-radius: 3px;
  padding: 1.25rem;
  line-height: 1.5;
  z-index: 5;
  opacity: 0.7;
  pointer-events: none;
  cursor: not-allowed;
  background-color: #3d4853;
`;

export namespace PriceButton {
  export interface Props {
    symbol: string;
    side: string;
    price: number;
    notional: number;
    execute: (tradeRequest: TradeRequest) => void;
    executing: boolean;
    url: string;
    movement?: string;
    isStale: boolean;
  }
}

export class PriceButton extends React.PureComponent<PriceButton.Props> {
  constructor(props: PriceButton.Props) {
    super(props);
    this.handleOnClick = this.handleOnClick.bind(this);
  }

  handleOnClick() {
    const tradeRequest: TradeRequest = {
      id: 1,
      symbol: this.props.symbol,
      broker: 'WEB',
      price: this.props.price,
      side: this.props.side,
      quantity: this.props.notional,
      url: this.props.url
    };
    this.props.execute(tradeRequest);
  }

  render() {
    const { side, price, executing, movement, isStale } = this.props;

    const classes = classNames({
      [style.oneWayPrice]: true,
      [style.buy]: side === 'buy',
      [style.sell]: side === 'sell',
      [style.nonTradeable]: price === 0,
      [style.executing]: executing === true
    });

    return (
      <div
        className="col-md-5"
        id="request-execution"
        onClick={this.handleOnClick}
      >
        <div className={`card   ${classes}`}>
          <h5 className="card-title">{side}</h5>
          {isStale && (
            <TileStaleStyle>
              <h5 className="card-title text-white bg-secondary">Stale</h5>
            </TileStaleStyle>
          )}

          <div className="card-body">
            <Price price={price} movement={movement} />
          </div>
        </div>
      </div>
    );
  }
}
