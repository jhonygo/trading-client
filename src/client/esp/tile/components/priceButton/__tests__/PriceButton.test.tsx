import React from 'react';
import { shallow } from 'enzyme';
import { PriceButton } from '../PriceButton';

describe('PriceButton', () => {
  it('should request execution', () => {
    // Given
    const execute = jest.fn();
    const price = shallow(
      <PriceButton
        symbol="USDGBP"
        price={1.18302}
        side="buy"
        notional={10000000}
        execute={execute}
        executing={false}
        url={''}
        isStale={false}
      />
    );

    // When
    price.find('#request-execution').simulate('click');

    // Then
    expect(execute).toHaveBeenCalledWith({
      id: 1,
      symbol: 'USDGBP',
      broker: 'WEB',
      price: 1.18302,
      side: 'buy',
      quantity: 10000000,
      url: ''
    });
  });

  it('should render buy price update', () => {
    // Given
    const price = shallow(
      <PriceButton
        symbol="USDGBP"
        price={1.18302}
        side="buy"
        notional={10000000}
        execute={jest.fn()}
        executing={false}
        url={''}
        isStale={false}
      />
    );

    // price
    price.setProps({ price: 1.43841 });

    // Then
    expect(price).toMatchSnapshot();
  });

  it('should render sell price update', () => {
    // Given
    const price = shallow(
      <PriceButton
        symbol="USDGBP"
        price={1.18302}
        side="sell"
        notional={10000000}
        execute={jest.fn()}
        executing={false}
        url={''}
        isStale={false}
      />
    );

    // When
    price.setProps({ price: 1.43841 });

    // Then
    expect(price).toMatchSnapshot();
  });

  it('should stale price', () => {
    // Given
    const price = shallow(
      <PriceButton
        symbol="USDGBP"
        price={1.18302}
        side="sell"
        notional={10000000}
        execute={jest.fn()}
        executing={false}
        url={''}
        isStale={true}
      />
    );

    // When
    price.setProps({ price: 1.43841 });

    // Then
    expect(price).toMatchSnapshot();
  });
});
