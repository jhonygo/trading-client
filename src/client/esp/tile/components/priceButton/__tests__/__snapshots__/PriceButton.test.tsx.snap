// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceButton should render buy price update 1`] = `ShallowWrapper {}`;

exports[`PriceButton should render sell price update 1`] = `ShallowWrapper {}`;

exports[`PriceButton should stale price 1`] = `ShallowWrapper {}`;
